pkg/scheduler/testing/workload_prep.go

```diff
@@ -65,3 +65,70 @@ func MakeNodesAndPods(pod *v1.Pod, existingPodsNum, allNodesNum, filteredNodesNu
 	}
 	return
 }
+
+// MakeNodesAndPodsForPodAffinity serves as a testing helper for Pod(Anti)Affinity feature.
+func MakeNodesAndPodsForPodAffinity(existingPodsNum, allNodesNum int) (existingPods []*v1.Pod, allNodes []*v1.Node) {
+	tpKeyToSizeMap := map[string]int{
+		"region": 3,
+		"zone":   10,
+		"node":   allNodesNum,
+	}
+	// build nodes to spread across all topology domains
+	for i := 0; i < allNodesNum; i++ {
+		nodeName := fmt.Sprintf("node%d", i)
+		nodeWrapper := MakeNode().Name(nodeName)
+		for tpKey, size := range tpKeyToSizeMap {
+			nodeWrapper = nodeWrapper.Label(tpKey, fmt.Sprintf("%s%d", tpKey, i%size))
+		}
+		allNodes = append(allNodes, nodeWrapper.Obj())
+	}
+
+	labels := []string{"foo", "bar", "baz"}
+	tpKeys := []string{"region", "zone", "node"}
+
+	// build pods
+	// we're going to apply PodAffinity with a matrix on `Left[*]` X `Right[*]` as showed below.
+	// left: [nil, hard PodAffinity, soft PodAffinity, hard and soft PodAffinity]
+	// right: [nil, hard PodAntiAffinity, soft PodAntiAffinity, hard and soft PodAntiAffinity]
+	left := []PodAffinityKind{
+		NilPodAffinity,
+		PodAffinityWithRequiredReq,
+		PodAffinityWithPreferredReq,
+		PodAffinityWithRequiredPreferredReq,
+	}
+	right := []PodAffinityKind{
+		NilPodAffinity,
+		PodAntiAffinityWithRequiredReq,
+		PodAntiAffinityWithPreferredReq,
+		PodAntiAffinityWithRequiredPreferredReq,
+	}
+
+	totalSize := len(left) * len(right)
+	for i := 0; i < existingPodsNum; i++ {
+		podWrapper := MakePod().Name(fmt.Sprintf("pod%d", i)).Node(fmt.Sprintf("node%d", i%allNodesNum))
+		label, tpKey := labels[i%len(labels)], tpKeys[i%len(tpKeys)]
+
+		affinityIdx := i % totalSize
+		// len(left) is equal to len(right)
+		leftIdx, rightIdx := affinityIdx/len(left), affinityIdx%len(left)
+		podWrapper = podWrapper.PodAffinityExists(label, tpKey, left[leftIdx])
+		podWrapper = podWrapper.PodAntiAffinityExists(label, tpKey, right[rightIdx])
+		existingPods = append(existingPods, podWrapper.Obj())
+	}
+
+	return
+}
+
+// MakeNodesAndPods serves as a testing helper to generate regular Nodes and Pods.
```

Comment:
What is regular? :)

Reply:
Regular means just plain cluster without using any advanced scheduling feature like Affinity and EvenPodsSpread.

Reply:
`MakeNodesAndPods serves a testing helper to generate Nodes and Pods that don't use any advanced scheduling features.`